### ISSUE TITLE
[BUG] Weight gradients with TransformerEngine v2.1 don't match those with TransformerEngine v1.12

### ISSUE THREAD

**Issue by okoge-kaz** [NVIDIA/TransformerEngine#1616]
**Describe the bug**

First, I found that the training loss with TE v2.1 doesn't match the one with TE v1.12, as shown below.
(using Megatron-LM mcore v0.11)

Training Loss (the higher one is run with TE v2.1, and the lower is run with TE v1.12)
![Image](https://github.com/user-attachments/assets/d80f21df-b45b-4f43-88d0-46f628ec116e)

Also, I found that the gradient norm doesn't match between the two versions.
![Image](https://github.com/user-attachments/assets/fcbd8c8a-d112-445c-aadc-d8e9e936b553)

To investigate which part is wrong, I added loggers to  [TELayerNormColumnParallelLinear's wgrad](https://github.com/NVIDIA/TransformerEngine/blob/v1.12/transformer_engine/pytorch/module/layernorm_linear.py#L651). The below is the result of logging of the wgrad. In Llama-3.1-8B training's first iteration, layer 32(= last layer)'s mlp linear_fc1 wgrad with TE v2.1 is the same as the one with TE 1.12, but layer 31's one is not. So, input_layernorm or self_attention's grad calculation should be wrong in TE 2.0 and later.
![Image](https://github.com/user-attachments/assets/46402c3c-096e-4a99-acaa-9836dd1db65a)

The model architecture I used to test is below. (TP=2)

```
      (decoder): TransformerBlock(
        (layers): ModuleList(
          (0-31): 32 x TransformerLayer(
            (input_layernorm): IdentityOp()
            (self_attention): SelfAttention(
              (core_attention): TEDotProductAttention(
                (flash_attention): FlashAttention()
                (fused_attention): FusedAttention()
                (unfused_attention): UnfusedDotProductAttention(
                  (scale_mask_softmax): FusedScaleMaskSoftmax()
                  (attention_dropout): Dropout(p=0.0, inplace=False)
                )
              )
              (linear_proj): TERowParallelLinear(in_features=2048, out_features=4096, bias=False, TP=2)
              (linear_qkv): TELayerNormColumnParallelLinear(in_features=4096, out_features=3072, bias=False, TP=2)
              (q_layernorm): IdentityOp()
              (k_layernorm): IdentityOp()
            )
            (pre_cross_attn_layernorm): IdentityOp()
            (cross_attention): IdentityOp()
            (cross_attn_bda): IdentityFuncOp()
            (pre_mlp_layernorm): IdentityOp()
            (mlp): MLP(
              (linear_fc1): TELayerNormColumnParallelLinear(in_features=4096, out_features=14336, bias=False, TP=2)
              (linear_fc2): TERowParallelLinear(in_features=7168, out_features=4096, bias=False, TP=2)
            )
          )
        )
        (final_layernorm): RMSNorm()
      )
```

**Steps/Code to reproduce bug**

I used [Megatron-LM mcore v0.11](https://github.com/NVIDIA/Megatron-LM/tree/v0.11.0) with the below script.
If tp-comm-overlap is disabled, the gradient mismatch is resolved. To reproduce this bug, please set TP >= 2 and enable tp-comm-overlap.
The versions of NGC PyTorch container are shown below.

- TE v2.1: nvcr.io/nvidia/pytorch:25.03-py3
- TE v1.12: nvcr.io/nvidia/pytorch:24.11-py3

```bash
HIDDEN_SIZE=4096
FFN_HIDDEN_SIZE=14336 # intermediate size (HuggingFace)
NUM_LAYERS=32
NUM_HEADS=32
NUM_KEY_VALUE_HEADS=8
SEQ_LENGTH=8192
MAX_POSITION_EMBEDDINGS=131072

# distributed settings
TENSOR_PARALLEL_SIZE=2
PIPELINE_PARALLEL_SIZE=1
CONTEXT_PARALLEL_SIZE=1
DATA_PARALLEL_SIZE=$((${NUM_GPUS} / (${TENSOR_PARALLEL_SIZE} * ${PIPELINE_PARALLEL_SIZE})))

PIPELINE_MODEL_CHUNKS=1
LAYERS_PER_VIRTUAL_PIPELINE_STAGE=$((${NUM_LAYERS} / ${PIPELINE_PARALLEL_SIZE} / ${PIPELINE_MODEL_CHUNKS}))

# training config
MICRO_BATCH_SIZE=2
GLOBAL_BATCH_SIZE=1024
TRAIN_STEPS=27500
LR_DECAY_ITERS=27500

LR=2.5E-5
MIN_LR=2.5E-6
LR_WARMUP_STEPS=1000
WEIGHT_DECAY=0.1
GRAD_CLIP=1


mpirun -np $NUM_GPUS \
  --npernode $NUM_GPU_PER_NODE \
  -hostfile $HOSTFILE_NAME \
  -x MASTER_ADDR=$MASTER_ADDR \
  -x MASTER_PORT=$MASTER_PORT \
  -x CUDA_DEVICE_MAX_CONNECTIONS=1 \
  -x NCCL_DEBUG=INFO \
  -x NCCL_IB_TIMEOUT=22 \
  ${DEBUG_ENVS} \
  ${NVTE_ARGS} \
  -bind-to none \
  python pretrain_gpt.py \
    --tensor-model-parallel-size ${TENSOR_PARALLEL_SIZE} \
    ${PIPELINE_ARGS} \
    --context-parallel-size ${CONTEXT_PARALLEL_SIZE} \
    --sequence-parallel \
    --use-distributed-optimizer \
    --overlap-param-gather \
    --overlap-grad-reduce \
    --tp-comm-overlap \
    --tp-comm-bootstrap-backend nccl \
    --distributed-timeout-minutes 30 \
    --num-layers ${NUM_LAYERS} \
    --hidden-size ${HIDDEN_SIZE} \
    --ffn-hidden-size ${FFN_HIDDEN_SIZE} \
    --num-attention-heads ${NUM_HEADS} \
    --group-query-attention \
    --num-query-groups ${NUM_KEY_VALUE_HEADS} \
    --seq-length ${SEQ_LENGTH} \
    --max-position-embeddings ${MAX_POSITION_EMBEDDINGS} \
    --micro-batch-size ${MICRO_BATCH_SIZE} \
    --global-batch-size ${GLOBAL_BATCH_SIZE} \
    --train-iters ${TRAIN_STEPS} \
    --tokenizer-type Llama3Tokenizer \
    --tokenizer-model ${TOKENIZER_MODEL} \
    ${CHECKPOINT_ARGS} \
    --save ${CHECKPOINT_SAVE_DIR} \
    --data-path ${TRAIN_DATA_PATH} \
    --split 990,10,0 \
    --distributed-backend nccl \
    --lr ${LR} \
    --min-lr ${MIN_LR} \
    --lr-decay-style cosine \
    --lr-decay-iters ${LR_DECAY_ITERS} \
    --weight-decay ${WEIGHT_DECAY} \
    --clip-grad ${GRAD_CLIP} \
    --lr-warmup-iters ${LR_WARMUP_STEPS} \
    --optimizer adam \
    --adam-beta1 0.9 \
    --adam-beta2 0.95 \
    --log-interval 1 \
    --save-interval 5000 \
    --ckpt-format "torch_dist" \
    --async-save \
    --eval-interval 500 \
    --eval-iters 10 \
    --bf16 \
    --untie-embeddings-and-output-weights \
    --no-position-embedding \
    --position-embedding-type rope \
    --rotary-base 500000.0 \
    --rotary-percent 1.0 \
    --use-rope-scaling \
    --disable-bias-linear \
    --use-mcore-models \
    --normalization RMSNorm \
    --norm-epsilon 1e-5 \
    --no-masked-softmax-fusion \
    --attention-dropout 0.0 \
    --hidden-dropout 0.0 \
    --swiglu \
    --attention-backend flash \
    --use-flash-attn \
    --attention-softmax-in-fp32 \
    --accumulate-allreduce-grads-in-fp32 \
    --transformer-impl "transformer_engine"
```

Training Loss with TE v2.1 (with/without tp-comm-overlap.)
![Image](https://github.com/user-attachments/assets/233d27ab-b03b-4cb6-a47c-960b46cfeeba)

Below are Training Losses with TE v2.1 and tp-comm-overlap, with TE v2.1 and without tp-comm-overlap, and with TE v1.12 and tp-comm-overlap.
( Training Loss with TE v2.1 and tp-comm-overlap disabled matches the one with TE v1.12 and tp-comm-overlap enabled. )
![Image](https://github.com/user-attachments/assets/d3c68b7c-7613-445c-9214-f47e905a3ae4)

| TE version | tp comm overlap | same as TE v1.12 |
| ---| ---|---|
| v2.1 | disabled | Yes. |
| v2.1 | enabled | No. |

**Expected behavior**

Training loss and gradient norms with TE v2.1 and tp-comm-overlap perfectly match those with TE v1.12. 

**Environment overview**

 - Environment location: Singularity(Container)
 - Method of Transformer Engine install: pre-installed (NGC container)
 - If method of install is [Docker], singularity build from .def file.

```bash
Bootstrap: docker
From: nvcr.io/nvidia/pytorch:25.03-py3

%post
    pip install --no-cache-dir wandb transformers
```

**Environment details**

NGC container tag information is shown below.

- TE v2.1: nvcr.io/nvidia/pytorch:25.03-py3
- TE v1.12: nvcr.io/nvidia/pytorch:24.11-py3

**Device details**
- GPU model: H200 141GB

**Additional context**

**Comment by BestJuly** [comment]
Hi @okoge-kaz, could you try the latest TE branch to see whether the bug still exists? 
We have found a bug in `tp_overlap` and have fixed it in [PR#1554](https://github.com/NVIDIA/TransformerEngine/pull/1554). The root cause is described [here](https://github.com/NVIDIA/TransformerEngine/pull/1554#discussion_r1993689321). We've observed similar divergence internally and after the fix, the loss curve goes back.

**Comment by okoge-kaz** [comment]
@BestJuly Thank you for your reply. I'll try.
 
> We have found a bug in tp_overlap and have fixed it in https://github.com/NVIDIA/TransformerEngine/pull/1554. The root cause is described https://github.com/NVIDIA/TransformerEngine/pull/1554#discussion_r1993689321. We've observed similar divergence internally and after the fix, the loss curve goes back

Thanks for sharing the detailed background!

**Comment by okoge-kaz** [comment]
The training loss divergence is resolved with the latest main, as shown below!
Thank you so much for TransformerEngine team's dedicated library maintenance work.

![Image](https://github.com/user-attachments/assets/2a0d0d1a-1a4d-438a-9684-0ef8bdb18054)
